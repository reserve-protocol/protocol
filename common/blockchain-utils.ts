import { BigNumber } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

// getChainId: Returns current chain Id
export const getChainId = async (hre: HardhatRuntimeEnvironment) => {
  let _chainId
  try {
    _chainId = await hre.network.provider.send('eth_chainId')
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...')
    _chainId = await hre.network.provider.send('net_version')
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`)
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString()
  }
  return _chainId
}
