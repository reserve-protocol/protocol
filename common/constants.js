const { BigNumber } = require("ethers");

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

const ONE_ETH= BigNumber.from("1000000000000000000");

const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1);
const MAX_UINT16 = (2**16) - 1;

const SCALE_FACTOR = 1e18;
const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString());

module.exports = {
    ZERO_ADDRESS,
    MAX_UINT256,
    MAX_UINT16,
    SCALE_FACTOR,
    BN_SCALE_FACTOR,
    ONE_ETH
};
