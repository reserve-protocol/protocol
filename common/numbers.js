const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}