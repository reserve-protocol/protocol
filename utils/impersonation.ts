import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

type ImpersonationFunction<T> = (signer: SignerWithAddress) => Promise<T>

/* whileImpersonating(address, f):

   Set up `signer` to be an ethers transaction signer that impersonates the account address
   `address`. In that context, call f(signer). `address` can be either a contract address or an
   external account, so you can use often this instead of building entire mock contracts.

   Example usage:

   await whileImpersonating(basketHandler.address, async (signer) => {
     await expect(rToken.connect(signer).setBasketsNeeded(fp('1'))
     .to.emit(rToken, 'BasketsNeededChanged')
   })

   This does the following:
   - Sets the basketHandler Eth balance to 2^256-1 (so it has plenty of gas)
   - Calls rToken.setBasketsNeeded _as_ the basketHandler contract,
   - Checks that that call emits the event 'BasketNeededChanged'
*/
export const whileImpersonating = async (
  hre: HardhatRuntimeEnvironment,
  address: string,
  f: ImpersonationFunction<void>
) => {
  // Set maximum ether balance at address
  await hre.network.provider.request({
    method: 'hardhat_setBalance',
    params: [address, '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff'],
  })
  await hre.network.provider.request({
    method: 'hardhat_impersonateAccount',
    params: [address],
  })
  const signer = await hre.ethers.getSigner(address)

  await f(signer)

  await hre.network.provider.request({
    method: 'hardhat_stopImpersonatingAccount',
    params: [address],
  })
  // If anyone ever needs it, we could make sure here that we set the balance at address back to
  // its original quantity...
}
