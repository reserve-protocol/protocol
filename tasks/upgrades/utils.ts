import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { IComponents } from '../../common/configuration'
import { IDeployments } from '../../scripts/deployment/common'
import { isValidContract } from '../../common/blockchain-utils'

export const validateDeployments = async (
  hre: HardhatRuntimeEnvironment,
  deployments: IDeployments,
  version: string
) => {
  // Check trading lib defined
  if (!deployments.tradingLib) {
    throw new Error(
      `Missing deployed TradingLib for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.tradingLib))) {
    throw new Error(`TradingLib contract not found in network ${hre.network.name}`)
  }

  // Check Main implementation is defined
  if (!deployments.implementations.main) {
    throw new Error(
      `Missing deployed Main implementation for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.implementations.main))) {
    throw new Error(`Main contract not found in network ${hre.network.name}`)
  }

  // Check all componet implementations are defined
  const componentsKeys: string[] = [
    'backingManager',
    'basketHandler',
    'broker',
    'distributor',
    'furnace',
    'rTokenTrader',
    'rsrTrader',
    'rToken',
    'stRSR',
  ]

  componentsKeys.forEach(async (keystr) => {
    if (!deployments.implementations.components[keystr as keyof IComponents]) {
      throw new Error(
        `Missing deployed ${keystr} implementation for version ${version} in network ${hre.network.name}`
      )
    } else if (
      !(await isValidContract(
        hre,
        deployments.implementations.components[keystr as keyof IComponents]
      ))
    ) {
      throw new Error(
        `Implementation ${keystr} contract not found for version ${version} in network ${hre.network.name}`
      )
    }
  })
}
