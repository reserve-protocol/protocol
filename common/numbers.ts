import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
