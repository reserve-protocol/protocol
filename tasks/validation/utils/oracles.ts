/* eslint-disable no-empty */
import { networkConfig } from '../../../common/configuration'
import { EACAggregatorProxyMock } from '@typechain/EACAggregatorProxyMock'
import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { BigNumber } from 'ethers'
import { AggregatorV3Interface } from '@typechain/index'
import { ONE_ADDRESS } from '../../../common/constants'

export const overrideOracle = async (
  hre: HardhatRuntimeEnvironment,
  oracleAddress: string
): Promise<EACAggregatorProxyMock> => {
  const oracle = await hre.ethers.getContractAt(
    'contracts/plugins/mocks/EACAggregatorProxyMock.sol:EACAggregatorProxy',
    oracleAddress
  )
  const aggregator = await oracle.aggregator()
  const accessController = await oracle.accessController()
  const initPrice = await oracle.latestRoundData()
  const mockOracleFactory = await hre.ethers.getContractFactory('EACAggregatorProxyMock')
  const mockOracle = await mockOracleFactory.deploy(aggregator, accessController, initPrice.answer)
  const bytecode = await hre.network.provider.send('eth_getCode', [mockOracle.address, 'latest'])
  await hre.network.provider.request({
    method: 'hardhat_setCode',
    params: [oracleAddress, bytecode],
  })
  return hre.ethers.getContractAt('EACAggregatorProxyMock', oracleAddress)
}

export const pushOraclesForward = async (
  hre: HardhatRuntimeEnvironment,
  rTokenAddress: string,
  extraAssets: string[] = []
) => {
  console.log(`🔃 Pushing Oracles forward for RToken: ${rTokenAddress}`)
  const rToken = await hre.ethers.getContractAt('RTokenP1', rTokenAddress)
  const main = await hre.ethers.getContractAt('IMain', await rToken.main())
  const assetRegistry = await hre.ethers.getContractAt(
    'AssetRegistryP1',
    await main.assetRegistry()
  )
  const registry = await assetRegistry.getRegistry()

  let addresses: string[] = [] // hacky way to ensure only unique updates to save time
  for (const asset of registry.assets) {
    addresses = await pushOracleForward(hre, asset, addresses)
  }

  for (const asset of extraAssets) {
    await pushOracleForward(hre, asset, addresses)
  }
}

export const pushOracleForward = async (
  hre: HardhatRuntimeEnvironment,
  asset: string,
  addresses: string[]
): Promise<string[]> => {
  // Need to handle all oracle cases, ie targetUnitChainlinkFeed, PoolTokens, etc
  const updateAnswer = async (chainlinkFeed: AggregatorV3Interface) => {
    if (addresses.indexOf(chainlinkFeed.address) != -1) return
    const initPrice = await chainlinkFeed.latestRoundData()
    const oracle = await overrideOracle(hre, chainlinkFeed.address)
    await oracle.updateAnswer(initPrice.answer)

    addresses.push(chainlinkFeed.address)
    console.log('✅ Feed Updated:', chainlinkFeed.address)
  }

  // chainlinkFeed
  try {
    const assetContract = await hre.ethers.getContractAt('TestIAsset', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContract.chainlinkFeed()
    )
    if (feed.address != ONE_ADDRESS) await updateAnswer(feed)
  } catch {
    // console.error('❌ chainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetUnitChainlinkFeed
  try {
    const assetContractNonFiat = await hre.ethers.getContractAt('NonFiatCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractNonFiat.targetUnitChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetUnitChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, refPerTokenChainlinkFeed
  try {
    const assetContractLido = await hre.ethers.getContractAt('L2LSDCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.exchangeRateChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ exchangeRateChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, refPerTokenChainlinkFeed
  try {
    const assetContractLido = await hre.ethers.getContractAt('L2LidoStakedEthCollateral', asset)
    let feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.targetPerRefChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.uoaPerTargetChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.refPerTokenChainlinkFeed()
    )
    await updateAnswer(feed)
    feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractLido.exchangeRateChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetPerRefChainlinkFeed, uoaPerTargetChainlinkFeed, or refPerTokenChainlinkFeed not found for:', asset, 'skipping...')
  }

  // targetPerTokChainlinkFeed
  try {
    const assetContractReth = await hre.ethers.getContractAt('RethCollateral', asset)
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      await assetContractReth.targetPerTokChainlinkFeed()
    )
    await updateAnswer(feed)
  } catch {
    // console.error('❌ targetPerTokChainlinkFeed not found for:', asset, 'skipping...')
  }

  // Dealing with nested RTokens
  // TODO do better

  // eUSDFRAXBP
  if (
    asset == '0x890FAa00C16EAD6AA76F18A1A7fe9C40838F9122' ||
    asset == '0x5cD176b58a6FdBAa1aEFD0921935a730C62f03Ac' ||
    asset == '0x994455cE66Fd984e2A0A0aca453e637810a8f032'
  ) {
    const feed = await hre.ethers.getContractAt(
      'AggregatorV3Interface',
      networkConfig['1'].chainlinkFeeds.FRAX!
    )
    await updateAnswer(feed)
    const eUSDAssetRegistry = await hre.ethers.getContractAt(
      'IAssetRegistry',
      '0x9B85aC04A09c8C813c37de9B3d563C2D3F936162'
    )
    const [, eUSDAssets] = await eUSDAssetRegistry.getRegistry()
    for (const eUSDAsset of eUSDAssets) {
      addresses = await pushOracleForward(hre, eUSDAsset, addresses) // recursion!
    }
  }

  return addresses
}

export const setOraclePrice = async (
  hre: HardhatRuntimeEnvironment,
  asset: string,
  value: BigNumber
) => {
  const assetContract = await hre.ethers.getContractAt('TestIAsset', asset)
  const realChainlinkFeed = await hre.ethers.getContractAt(
    'AggregatorV3Interface',
    await assetContract.chainlinkFeed()
  )
  const oracle = await overrideOracle(hre, realChainlinkFeed.address)
  await oracle.updateAnswer(value)
}
