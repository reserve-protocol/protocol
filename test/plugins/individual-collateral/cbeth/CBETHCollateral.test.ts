import collateralTests from '../collateralTests'
import { CollateralFixtureContext, CollateralOpts, MintCollateralFunc } from '../pluginTestTypes'
import {
  CBETH_ETH_PRICE_FEED,
  CB_ETH,
  CB_ETH_ORACLE,
  DEFAULT_THRESHOLD,
  DELAY_UNTIL_DEFAULT,
  ETH_USD_PRICE_FEED,
  MAX_TRADE_VOL,
  ORACLE_ERROR,
  ORACLE_TIMEOUT,
  PRICE_TIMEOUT,
} from './constants'
import { pushOracleForward } from '../../../utils/oracles'
import { BigNumber, BigNumberish, ContractFactory } from 'ethers'
import { bn, fp } from '#/common/numbers'
import { TestICollateral } from '@typechain/TestICollateral'
import { ethers } from 'hardhat'
import { expect } from 'chai'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { MockV3Aggregator } from '@typechain/MockV3Aggregator'
import { ICBEth, ERC20Mock, MockV3Aggregator__factory } from '@typechain/index'
import { mintCBETH, resetFork } from './helpers'
import { whileImpersonating } from '#/utils/impersonation'
import hre from 'hardhat'

interface CbEthCollateralFixtureContext extends CollateralFixtureContext {
  cbETH: ICBEth
  targetPerTokChainlinkFeed: MockV3Aggregator
}

interface CbEthCollateralOpts extends CollateralOpts {
  targetPerTokChainlinkFeed?: string
  targetPerTokChainlinkTimeout?: BigNumberish
}

export const deployCollateral = async (
  opts: CbEthCollateralOpts = {}
): Promise<TestICollateral> => {
  opts = { ...defaultCBEthCollateralOpts, ...opts }

  const CBETHCollateralFactory: ContractFactory = await ethers.getContractFactory('CBEthCollateral')

  const collateral = <TestICollateral>await CBETHCollateralFactory.deploy(
    {
      erc20: opts.erc20,
      targetName: opts.targetName,
      priceTimeout: opts.priceTimeout,
      chainlinkFeed: opts.chainlinkFeed,
      oracleError: opts.oracleError,
      oracleTimeout: opts.oracleTimeout,
      maxTradeVolume: opts.maxTradeVolume,
      defaultThreshold: opts.defaultThreshold,
      delayUntilDefault: opts.delayUntilDefault,
    },
    opts.revenueHiding,
    opts.targetPerTokChainlinkFeed ?? CBETH_ETH_PRICE_FEED,
    opts.targetPerTokChainlinkTimeout ?? ORACLE_TIMEOUT,
    { gasLimit: 2000000000 }
  )
  await collateral.deployed()

  // Push forward chainlink feeds
  await pushOracleForward(opts.chainlinkFeed!)
  await pushOracleForward(opts.targetPerTokChainlinkFeed ?? CBETH_ETH_PRICE_FEED)

  await expect(collateral.refresh())

  return collateral
}

const chainlinkDefaultAnswer = bn('1600e8')
const targetPerTokChainlinkDefaultAnswer = fp('1.04027709')

type Fixture<T> = () => Promise<T>

const makeCollateralFixtureContext = (
  alice: SignerWithAddress,
  opts: CbEthCollateralOpts = {}
): Fixture<CbEthCollateralFixtureContext> => {
  const collateralOpts = { ...defaultCBEthCollateralOpts, ...opts }

  const makeCollateralFixtureContext = async () => {
    const MockV3AggregatorFactory = <MockV3Aggregator__factory>(
      await ethers.getContractFactory('MockV3Aggregator')
    )

    const chainlinkFeed = <MockV3Aggregator>(
      await MockV3AggregatorFactory.deploy(8, chainlinkDefaultAnswer)
    )
    collateralOpts.chainlinkFeed = chainlinkFeed.address
    const targetPerTokChainlinkFeed = <MockV3Aggregator>(
      await MockV3AggregatorFactory.deploy(18, targetPerTokChainlinkDefaultAnswer)
    )

    collateralOpts.targetPerTokChainlinkFeed = targetPerTokChainlinkFeed.address
    collateralOpts.targetPerTokChainlinkTimeout = ORACLE_TIMEOUT

    const cbETH = (await ethers.getContractAt('ICBEth', CB_ETH)) as unknown as ICBEth
    const collateral = await deployCollateral(collateralOpts)

    return {
      alice,
      collateral,
      chainlinkFeed,
      targetPerTokChainlinkFeed,
      cbETH,
      tok: cbETH as unknown as ERC20Mock,
    }
  }

  return makeCollateralFixtureContext
}
/*
  Define helper functions
*/

const mintCollateralTo: MintCollateralFunc<CbEthCollateralFixtureContext> = async (
  ctx: CbEthCollateralFixtureContext,
  amount: BigNumberish,
  user: SignerWithAddress,
  recipient: string
) => {
  await mintCBETH(amount, recipient)
}

const changeTargetPerRef = async (ctx: CbEthCollateralFixtureContext, percentChange: BigNumber) => {
  // We leave the actual refPerTok exchange where it is and just change {target/tok}
  {
    const lastRound = await ctx.targetPerTokChainlinkFeed.latestRoundData()
    const nextAnswer = lastRound.answer.add(lastRound.answer.mul(percentChange).div(100))
    await ctx.targetPerTokChainlinkFeed.updateAnswer(nextAnswer)
  }
}

const reduceTargetPerRef = async (
  ctx: CbEthCollateralFixtureContext,
  pctDecrease: BigNumberish
) => {
  await changeTargetPerRef(ctx, bn(pctDecrease).mul(-1))
}

const increaseTargetPerRef = async (
  ctx: CbEthCollateralFixtureContext,
  pctDecrease: BigNumberish
) => {
  await changeTargetPerRef(ctx, bn(pctDecrease))
}

const changeRefPerTok = async (ctx: CbEthCollateralFixtureContext, percentChange: BigNumber) => {
  await whileImpersonating(hre, CB_ETH_ORACLE, async (oracleSigner) => {
    const rate = await ctx.cbETH.exchangeRate()
    await ctx.cbETH
      .connect(oracleSigner)
      .updateExchangeRate(rate.add(rate.mul(percentChange).div(bn('100'))))
    {
      const lastRound = await ctx.targetPerTokChainlinkFeed.latestRoundData()
      const nextAnswer = lastRound.answer.add(lastRound.answer.mul(percentChange).div(100))
      await ctx.targetPerTokChainlinkFeed.updateAnswer(nextAnswer)
    }

    {
      const lastRound = await ctx.chainlinkFeed.latestRoundData()
      const nextAnswer = lastRound.answer.add(lastRound.answer.mul(percentChange).div(100))
      await ctx.chainlinkFeed.updateAnswer(nextAnswer)
    }
  })
}

const reduceRefPerTok = async (ctx: CbEthCollateralFixtureContext, pctDecrease: BigNumberish) => {
  await changeRefPerTok(ctx, bn(pctDecrease).mul(-1))
}

const increaseRefPerTok = async (ctx: CbEthCollateralFixtureContext, pctIncrease: BigNumberish) => {
  await changeRefPerTok(ctx, bn(pctIncrease))
}
const getExpectedPrice = async (ctx: CbEthCollateralFixtureContext): Promise<BigNumber> => {
  const clData = await ctx.chainlinkFeed.latestRoundData()
  const clDecimals = await ctx.chainlinkFeed.decimals()

  const clRptData = await ctx.targetPerTokChainlinkFeed.latestRoundData()
  const clRptDecimals = await ctx.targetPerTokChainlinkFeed.decimals()

  return clData.answer
    .mul(bn(10).pow(18 - clDecimals))
    .mul(clRptData.answer.mul(bn(10).pow(18 - clRptDecimals)))
    .div(fp('1'))
}

/*
  Define collateral-specific tests
*/

// eslint-disable-next-line @typescript-eslint/no-empty-function
const collateralSpecificConstructorTests = () => {
  it('does not allow missing targetPerTok chainlink feed', async () => {
    await expect(
      deployCollateral({ targetPerTokChainlinkFeed: ethers.constants.AddressZero })
    ).to.be.revertedWith('missing targetPerTok feed')
  })

  it('does not allow targetPerTok oracle timeout at 0', async () => {
    await expect(deployCollateral({ targetPerTokChainlinkTimeout: 0 })).to.be.revertedWith(
      'targetPerTokChainlinkTimeout zero'
    )
  })
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const collateralSpecificStatusTests = () => {}
// eslint-disable-next-line @typescript-eslint/no-empty-function
const beforeEachRewardsTest = async () => {}

export const defaultCBEthCollateralOpts: CollateralOpts = {
  erc20: CB_ETH,
  targetName: ethers.utils.formatBytes32String('ETH'),
  priceTimeout: PRICE_TIMEOUT,
  chainlinkFeed: ETH_USD_PRICE_FEED,
  oracleTimeout: ORACLE_TIMEOUT,
  oracleError: ORACLE_ERROR,
  maxTradeVolume: MAX_TRADE_VOL,
  defaultThreshold: DEFAULT_THRESHOLD,
  delayUntilDefault: DELAY_UNTIL_DEFAULT,
  revenueHiding: fp('0'),
}

/*
  Run the test suite
*/

const opts = {
  deployCollateral,
  collateralSpecificConstructorTests,
  collateralSpecificStatusTests,
  beforeEachRewardsTest,
  makeCollateralFixtureContext,
  mintCollateralTo,
  reduceTargetPerRef,
  increaseTargetPerRef,
  reduceRefPerTok,
  increaseRefPerTok,
  getExpectedPrice,
  itClaimsRewards: it.skip,
  itChecksTargetPerRefDefault: it,
  itChecksTargetPerRefDefaultUp: it,
  itChecksRefPerTokDefault: it,
  itChecksPriceChanges: it,
  itHasRevenueHiding: it,
  itChecksNonZeroDefaultThreshold: it,
  resetFork,
  collateralName: 'CBEthCollateral',
  chainlinkDefaultAnswer,
  itIsPricedByPeg: true,
}

collateralTests(opts)
