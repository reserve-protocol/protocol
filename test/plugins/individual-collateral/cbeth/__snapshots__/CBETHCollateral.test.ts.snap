// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting ERC20 transfer 1`] = `65479`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting ERC20 transfer 2`] = `48379`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after full price timeout 1`] = `59813`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after full price timeout 2`] = `55344`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after hard default 1`] = `98295`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after hard default 2`] = `90612`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after oracle timeout 1`] = `59813`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after oracle timeout 2`] = `55344`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after soft default 1`] = `55003`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() after soft default 2`] = `55003`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() during SOUND 1`] = `90609`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() during SOUND 2`] = `90609`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() during soft default 1`] = `98160`;

exports[`Collateral: CBEthCollateral collateral functionality Gas Reporting refresh() during soft default 2`] = `90891`;
